apps/labs/posts/array-api-support-scikit-learn.md

```diff
@@ -0,0 +1,144 @@
+
+```python
+import sklearn
+import torch
+from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
+
+X_torch_cpu, y_torch_cpu = torch.asarray(X_np), torch.asarray(y_np)
+
+with sklearn.config_context(array_api_dispatch=True):
+    lda_torch_cpu = LinearDiscriminantAnalysis()
+    lda_torch_cpu.fit(X_torch_cpu, y_torch_cpu)
+
+    predictions = lda_torch_cpu.predict(X_torch_cpu)
+
+print(type(predictions))
+# <class 'torch.Tensor'>
+```
+
+Note how the estimator's output returns an array from the input's array library. This following example uses the global configuration and PyTorch Tensors on GPUs:
+
+```python
+import sklearn
+import torch
+
+sklearn.set_config(array_api_dispatch=True)
+
+X_torch_cuda = torch.asarray(X_np, device="cuda")
+y_torch_cuda = torch.asarray(y_np, device="cuda")
+
+lda_torch_cuda = LinearDiscriminantAnalysis()
+lda_torch_cuda.fit(X_torch_cuda, y_torch_cuda)
+predictions = lda_torch_cuda.predict(X_torch_cuda)
+
+print(predictions.device)
+# cuda:0
+```
+
+A common machine learning use case is to train a model on a GPU and then transfer it to a CPU for deployment. Scikit-learn provides a private utility function to handle this device movement:
+
+```python
+from sklearn.utils._array_api import _estimator_with_converted_arrays
+
+tensor_to_ndarray = lambda array : array.cpu().numpy()
+lda_np = _estimator_with_converted_arrays(lda_torch_cuda, tensor_to_ndarray)
+X_trans = lda_np.transform(X_np)
+
+print(type(X_trans))
+# <class 'numpy.ndarray'>
+```
+
+You can learn more about Scikit-learn's Array API support in their [documentation](https://scikit-learn.org/dev/modules/array_api.html).
+
+## Challenges
+
+Adopting the Array API standard in scikit-learn was not a straightforward task and required us to overcome some challenges. The significant challenges fall into two broad categories:
+
+- Array API Standard is a subset of NumPy's API
+- Compiled code that only run on CPUs that was written in C, C++, or Cython.
+
+### Array API Standard is a subset of NumPy's API
+
+NumPy's API is extensive and includes a vast amount of operations. By design, the Array API standard is a subset of the NumPy API. For functionality to be included in the Array API standard, it must be implemented by most array libraries and widely used. Scikit-learn's codebase was initially written to use NumPy's API. In order to adopt the Array API, we had to rewrite some NumPy functions in terms of Array API functions. For example, `nanmin` is not a part of the Array API standard, so we were required to implement it:
+
+```python
+def _nanmin(X, axis=None):
+    xp = get_array_namespace(X)
+    if _is_numpy_namespace(xp):
+        return xp.asarray(numpy.nanmin(X, axis=axis))
+    else:
+        # Implements nanmin in terms of the Array API standard
+        mask = xp.isnan(X)
+        X = xp.min(..., X), axis=axis)
```

Comment:
Maybe I am missing something here, but how is `mask` being used in the current snippet?

Reply:
I did not want to complicate the snippet more with the implementation details. But it is a good point to at least use the `mask`. I update the PR: [`811958c` (#765)](https://github.com/Quansight/Quansight-website/pull/765/commits/811958c988c6a66202a310c5cff006de0a2f132b) with your suggestion.